test/zdtm/static/change_mnt_context.c

```diff
@@ -0,0 +1,60 @@
+#include <sys/mount.h>
+#include <sys/stat.h>
+
+#include "zdtmtst.h"
+
+const char *test_doc = "Check that mounts can change context on restore";
+const char *test_author = "Adrian Reber <areber@redhat.com>";
+
+char *source = "change_mount_context.test";
+char *original_context = "context=\"system_u:object_r:container_file_t:s0:c82,c137\"";
+char *new_context = "context=\"system_u:object_r:container_file_t:s0:c204,c495\"";
+
+int main(int argc, char **argv)
+{
+	char *dname = "/tmp/change_mount_context.XXXXXX";
+	char line[1024];
+	char opts[1024];
+	int ret = -1;
+	FILE *mi;
+
+	mkdir(dname, 755);
+	if (mount(source, dname, "tmpfs", 0, original_context)) {
+		pr_perror("mount");
+		return 1;
+	}
+
+	test_init(argc, argv);
+
+	test_daemon();
+	test_waitsig();
+
+	mi = fopen("/proc/self/mountinfo", "r");
+	if (mi == NULL) {
+		fail("No mountinfo file");
+		goto out;
+	}
+
+	while (fgets(line, sizeof(line), mi)) {
+		int result;
+
+		if (!strstr(line, source))
+			continue;
+
+		result = sscanf(line, "%*d %*d %*d:%*d %*s %*s %*s - %*s %*s %s", opts);
```

Comment:
Should use `strstr(line, " - ")` as before `-` we can have variable number of arguments.

Reply:
Brief check shows that you haven't check the problem still.

I've just took your parse string and tried parsing one simple mount line in mountinfo...

```
[snorch@fedora tmp]$ cat test.c 
#include <stdio.h>
#include <string.h>

#define MNTINFO_EX "96 90 259:1 / /boot/efi rw,relatime shared:48 - vfat /dev/nvme0n1p1 rw,fmask=0077,dmask=0077,codepage=437,iocharset=ascii,shortname=winnt,errors=remount-ro"

int main()
{
	char buf[1024];
	int ret;

	ret = sscanf(MNTINFO_EX, "%*d %*d %*d:%*d %*s %*s %*s - %*s %*s %s", buf);
	if (ret != 1) {
		printf("Failed sscanf: %d\n", ret);
		return 1;
	}

	printf("Parsed :\"%s\"\n", buf);
	return 0;
}
[snorch@fedora tmp]$ gcc -o test test.c
[snorch@fedora tmp]$ ./test
Failed sscanf: 0
```

Reply:
Any recommendations how to make the parsing better?

I added
```

		if (!strstr(line, " - "))
			continue;
```

a couple of days ago and I am only looking at line that contain source of the mount `change_mount_context.test`. Any check I can add to make sure it works?

Reply:
> Should use strstr(line, " - ") as before - we can have variable number of arguments.

@Snorch do you mean something like this:

diff --git a/test/zdtm/static/change_mnt_context.c b/test/zdtm/static/change_mnt_context.c
index 4d2bf7ccd..6d436014b 100644
--- a/test/zdtm/static/change_mnt_context.c
+++ b/test/zdtm/static/change_mnt_context.c
@@ -37,14 +37,16 @@ int main(int argc, char **argv)
 
        while (fgets(line, sizeof(line), mi)) {
                int result;
+               char *pos;
 
                if (!strstr(line, source))
                        continue;
 
-               if (!strstr(line, " - "))
+               pos = strstr(line, " - ");
+               if (!pos)
                        continue;
 
-               result = sscanf(line, "%*d %*d %*d:%*d %*s %*s %*s - %*s %*s %s", opts);
+               result = sscanf(pos, " - %*s %*s %s", opts);
                if (result != 1) {
                        fail("Not able to sscanf line from mountinfo");
                        goto out;

Reply:
Thanks @avagin for the example.